tservice/types/src/seal.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Types used for sealing and unsealing of secrets
+
+use mc_sgx_core_types::KeyRequest;
+
+/// AES GCM(Galois/Counter mode) Data
+///
+/// Wraps up a `&[u8]` since [mc-sgx-tservice-sys-types::sgx_aes_gcm_data_t] is
+/// a dynamically sized type
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct AesGcmData<'a> {
+    bytes: &'a [u8],
+}
+
+impl<'a> From<&'a [u8]> for AesGcmData<'a> {
+    fn from(bytes: &'a [u8]) -> Self {
+        Self { bytes }
+    }
+}
+
+impl<'a> AesGcmData<'a> {
+    const PAYLOAD_OFFSET: usize = 32;
+
+    /// The size of the payload (encrypted data + mac text)
+    fn payload_size(&self) -> usize {
+        let size = u32::from_le_bytes(
+            self.bytes[..4]
+                .try_into()
+                .expect("Failed to extract `payload_size`"),
+        );
+        size as usize
+    }
+
+    /// The GMAC tag of the payload
+    pub fn gmac_tag(&self) -> [u8; 16] {
+        self.bytes[16..32]
+            .try_into()
+            .expect("Failed to extract `payload_tag`")
+    }
+
+    /// The payload.  This includes the encrypted data and the MAC text
+    pub fn payload(&self) -> &'a [u8] {
+        let end = self.payload_size() + Self::PAYLOAD_OFFSET;
+        self.bytes
+            .get(Self::PAYLOAD_OFFSET..end)
+            .expect("AesGcmData not large enough to contain the payload.")
+    }
+}
+
+/// Sealed data
+///
+/// Wraps up a `&[u8]` since [mc-sgx-tservice-sys-types::sgx_sealed_data_t]
+/// is a dynamically sized type
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct SealedData<'a> {
+    bytes: &'a [u8],
+}
+
+impl<'a> SealedData<'a> {
+    /// The key request used to seal the data
+    pub fn key_request(&self) -> KeyRequest {
+        let bytes: [u8; 512] = self.bytes[..512]
+            .try_into()
+            .expect("SealedData bytes aren't big enough to hold `key_request`");
+        KeyRequest::from(&bytes)
+    }
+
+    // The offset within the `aes_data` for where the MAC starts.
+    // None when there is no MAC data.
+    fn mac_offset(&self) -> Option<usize> {
```

Comment:
probably a mix of bad naming on my part and questionable naming in the C api, https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_tseal.h#L59

They seem to use "MAC text" a lot, and then in  a few places use [aad](https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_tseal.h#L195)
My vote would be to deviate from the API and use AAD everywhere with something  in the readme or similar to let people know the intended disambiguation.

I would also like to standardize on `size` vs `len`.  The API uses both terms.  It may just be me and my history or it may be general C and embedded practice, but my understanding is `size` == number of bytes and `len` == number of items.  The type is a uint8 or u8, but with them throwing around terms like `text` it would be unfortunate for someone to mix that up with a UTF-8 String

Reply:
function removed now